### DISCUSSION TITLE
Preprocessing the whole context with an external command

### DISCUSSION THREAD

**Discussion by janbuchar** [orhun/git-cliff#773] [Ideas]
Has support for preprocessing the context object and potentially adding new fields and then piping this back into the templating system been considered?

I have encountered several use cases that could be resolved by other means, but just having one preprocessor script could be easier to understand.

**Comment by orhun** [comment]
Hello, I haven't considered adding this before but I'm definitely interested in hearing about your use case more. In which cases do you think this extension point can be useful?

Also, there is #657 which is semi-related to this.

**Reply by janbuchar** [reply]
Well, I have two use cases - one is adding a field with numbers of issues closed by a commit to the context so that I could print it in the changelog and leave the presentation details in the template. Currently I'm trying to hack around this by postprocessing the changelog.

My other use case is having a changelog-errata file that could be used to adjust the commit messages in case we mess up and want to fix the changelog. This can be achieved quite neatly with preprocessors, I believe.

However, I think that allowing transformations on the context would be intuitive and maybe even powerful enough to replace the more limitted preprocessors, postprocessors and link processors. 

The linked issue sounds like a feasible way to do this.

**Reply by muja** [reply]
> one is adding a field with numbers of issues closed by a commit

You can already do this without any hacks by putting those issue numbers into trailers/footers and then accessing them via the `commit.footers`

**Reply by janbuchar** [reply]
Fair enough. I can't rewrite public history though, and there's no way I can convince my team to put that info in there.

**Reply by muja** [reply]
The team doesn't need to be convinced IMHO. They just need to be educated that if they want the issues to appear in the release notes, they have to put them in the footer.

**Reply by janbuchar** [reply]
You're not wrong. Still, I just want to generate better changelog though, not stir up bikeshedding about git best practices. And when https://github.com/orhun/git-cliff/pull/784 is released, I'll be able to do it super cleanly.

**Reply by orhun** [reply]
#784 is now released :) https://git-cliff.org/blog/2.5.0
